core/drivers/clk/sam/at91_master.c

```diff
@@ -179,3 +182,110 @@ const struct clk_master_layout at91sam9x5_master_layout = {
 	.pres_shift = 4,
 	.offset = AT91_PMC_MCKR,
 };
+
+static size_t clk_sama7g5_master_get_parent(struct clk *hw)
+{
+	struct clk_master *master = hw->priv;
+	size_t i = 0;
+
+	for (i = 0; i < hw->num_parents; i++)
+		if (master->mux_table[i] == master->parent)
+			return i;
+
+	panic("Can't get correct parent of clock");
+}
+
+static TEE_Result clk_sama7g5_master_set_parent(struct clk *hw, size_t index)
+{
+	struct clk_master *master = hw->priv;
+
+	if (index >= hw->num_parents)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	master->parent = master->mux_table[index];
+	return TEE_SUCCESS;
+}
+
+static TEE_Result clk_sama7g5_master_set_rate(struct clk *hw,
+					      unsigned long rate,
+					      unsigned long parent_rate)
+{
+	struct clk_master *master = hw->priv;
+	unsigned long div = 0;
+
+	div = UDIV_ROUND_NEAREST(parent_rate, rate);
+	if ((div > (1 << (MASTER_PRES_MAX - 1))) || (div & (div - 1)))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (div == 3)
+		div = MASTER_PRES_MAX;
+	else if (div)
+		div = ffs(div) - 1;
+	master->div = div;
+	return TEE_SUCCESS;
+}
+
+static unsigned long clk_sama7g5_master_get_rate(struct clk *hw,
+						 unsigned long parent_rate)
+{
+	struct clk_master *master = hw->priv;
+	unsigned long rate = parent_rate >> master->div;
+
+	if (master->div == 7)
+		rate = parent_rate / 3;
+
+	return rate;
+}
+
+static const struct clk_ops sama7g5_master_ops = {
+	.set_rate = clk_sama7g5_master_set_rate,
+	.get_rate = clk_sama7g5_master_get_rate,
+	.get_parent = clk_sama7g5_master_get_parent,
+	.set_parent = clk_sama7g5_master_set_parent,
+};
+
+struct clk *at91_clk_sama7g5_register_master(struct pmc_data *pmc,
+					     const char *name,
+					     int num_parents,
+					     struct clk **parent,
+					     uint32_t *mux_table,
+					     uint8_t id,
+					     int chg_pid)
+{
+	struct clk_master *master = NULL;
+	struct clk *hw = NULL;
+	unsigned int val = 0;
+
+	if (!name || !num_parents || !parent || !mux_table ||
+	    id > MASTER_MAX_ID)
+		return NULL;
+
+	master = calloc(1, sizeof(*master));
+	if (!master)
+		return NULL;
+
+	hw = clk_alloc(name, &sama7g5_master_ops, parent, num_parents);
+	if (!hw) {
+		free(master);
+		return NULL;
+	}
+
+	hw->priv = master;
+	master->base = pmc->base;
+	master->id = id;
+	master->chg_pid = chg_pid;
+	master->mux_table = mux_table;
+
+	io_write32(master->base + AT91_PMC_MCR_V2, master->id);
```

Comment:
build fails on `AT91_PMC_MCR_V2*` macros not being defined.

Reply:
The definitions of these macros are added in #6630, you can find them in [`core/drivers/clk/sam/at91_pmc.h`](https://github.com/OP-TEE/optee_os/pull/6630/files).
It's better to integrate #6630 before this PR, thanks.